### DISCUSSION TITLE
Filter design pattern

### DISCUSSION THREAD

**Discussion by yongjie1** [twilio-labs/paste#1979] [Office hours] [answered]
In the Twilio intelligence product, users (e.g. the call center manager ) need to be able to filter a list of calls and find the ones that are important or relevant to them. Users can filter the calls by agent name, customers name, operator name (intelligence insights) and find the calls that are important to them.

The most complicated part is the filtering by operator name, as it has 2 levels of filtering: 
- level 1 is filtering by the operator name
- level 2 is filtering by the label name within the selected operator

And users can also filter by multiple operators and each of the operators could has multiple labels.

Here are a few different concepts that we've explored:
V1 - Hide the filter in the side panel:
![image](https://user-images.githubusercontent.com/91665462/139161677-1fefa2b9-a62c-4849-86a6-b2fc5e7060cd.png)

![image](https://user-images.githubusercontent.com/91665462/139127945-cbb1af9d-0a4d-4845-b0d3-516a47493188.png)

V2 Show all the filters at once
![image](https://user-images.githubusercontent.com/91665462/139128116-6972814a-965b-4525-a002-90da23ae84b6.png)

V3 Show some important filters and hide some additional filters in an accordion 
![image](https://user-images.githubusercontent.com/91665462/139161091-ebfa581f-e404-44f4-8bcd-85bd09bdc975.png)

**Comment by gloriliale** [comment]
Hi @yongjie1, wanted to add the related discussions you mentioned for context:
[1601: Filter Pattern contribution thread](https://github.com/twilio-labs/paste/discussions/1601)
[1702: Designing complex Filters for data](https://github.com/twilio-labs/paste/discussions/1702)
[1715: Search / filter on a data table](https://github.com/twilio-labs/paste/discussions/1715)
[1366: Combobox Group](https://github.com/twilio-labs/paste/discussions/1366)
[865: Table Filters for bulk action](https://github.com/twilio-labs/paste/discussions/865)
[926: Filter data table (filter vs. search)](https://github.com/twilio-labs/paste/discussions/926 )

**Reply by kelseycavitt** [reply]
Hi @yongjie1 !
Another useful link is Corinne's first draft of the [Filter Group Pattern Spec](https://docs.google.com/document/d/1w7_BIeFNlg3b8Nm_r-2WlqTR7MQyhcI5vFmF_rM-pnQ/edit?usp=sharing)

**Comment by kelseycavitt** [comment]
Hi @yongjie!
Thanks for a great discussion yesterday with yourself and Qi Feng.

Here’s a link to the [Office hours discussion recording](https://www.loom.com/share/646f0ebb86f34fc8ad44de3f29a4884b).

Here’s a recap to what we discussed:

1. We are encouraging you to pursue the use of the side panel / drawer container for your filtering needs, as there are several possible filters that can be applied in your context/use case, specifically with the multiple levels for operators. With pursuing this approach, we are hoping to have this kind of component on our roadmap in the future, so when it’s available, it’s expected that the component would need to be switched out for the new one for Paste. This kind of pattern for filtering exists currently in the Flex product as well.
2. We both discussed looking into best practices from a UX and accessibility perspective for displaying selected filters (e.g., exposing them outside of the drawer, or just keeping them in the drawer). I’ll be working on some research/suggestions these next few days!
3. For your combobox needs, we suggest using the [Autocomplete combobox.](https://paste.twilio.design/components/combobox#autocomplete-combobox )
4. If you’re using disclosures, specifically for the Agent name that uses a checkbox group, ensure the checkbox group has a label, as the disclosure label only gives a label to the disclosure, not the form elements inside it.
5. For the operator filter, we suggest moving the “add operator filter” action to the top, as a screenreader could miss that action when being read, a user could also be confused and it could get moved below the fold when adding more operator filters. Additionally, the action should be a [link-style button](https://paste.twilio.design/components/button#link-style-button).
6. As discussed, let’s do another design review pass once you’re ready to help ensure the best usage of some design tokens, namely the spacing and possibly box shadow for the drawer/side panel.

**Reply by yongjie1** [reply]
@kelseycavitt Thanks for the summary. It was great to chat with you and the design system team about the filter experience within our product.

For the point 3, we'll need autocomplete combobox for multi selection. I've seen [this example]( https://paste.twilio.design/primitives/combobox-primitive/#multiselect-combobox-example). 
However, as a users if I need select multiple options, I have to select one, then click on the input box again and select another one, and so on. This isn't efficient for users. Is it possible to do something like this for multi selection(See screen)? 
![image](https://user-images.githubusercontent.com/91665462/139727445-b7a71f9a-9a32-4e56-ba68-ba3c6ea366a1.png)

For the point 4, if the form inside the disclosure is very simple, do we still need another label name for the form (see the screen example)? It feels repetitive. I think what you have suggested make sense if we have a form with different sections and it should have section title but I'm not sure if it works for our use case.
![image](https://user-images.githubusercontent.com/91665462/139728217-56e387cb-bdb8-452b-9f6c-e2df5c853f14.png)

**Comment by kelseycavitt** [comment]
Hi @yongjie1!

For the mulit-select combobox needs, yes! You can use the Combobox primitive to do what you want to do in your design. 

As for the form label question, no, you don't need an extra form label here, as the text inputs themselves have labels. It was just for that example for the checkbox group, as that didn't seem to have a visible label in your design.

Let me know if this makes sense! I'm also still looking into some best filtering practices for displaying selected filters, hoping to have something to you soon!

**Reply by yongjie1** [reply]
@kelseycavitt Thanks. 
 
I've found some examples: [Shopify](https://polaris.shopify.com/components/lists-and-tables/filters) 
![image](https://user-images.githubusercontent.com/91665462/139907305-b3a5e272-6e6e-4c8a-a7ef-d32ac4e5add7.png)

Another example from AirB&B
![image](https://user-images.githubusercontent.com/91665462/139907927-f801d2d1-fd08-48f9-8ae6-60bea88577d1.png)

I'm still try to find out if there is any good articles talking about filters and the best practices. Let me know when you found anything useful.

**Comment by kelseycavitt** [comment]
@yongjie1 ,
Great examples! I think I found a couple of useful articles that point to extensive user research on best practices for filtering. 

1. [How to design applied filters ](https://baymard.com/blog/how-to-design-applied-filters)(Baymard)
2. [Crafting a kickass filtering ux ](https://uxdesign.cc/crafting-a-kickass-filtering-ux-beea1798d64b)(Medium)

It seems that user research has shown that the UX is better off with showing both applied filters in their original position, AND in a separate ‘Applied Filters’ section. I think that with the designs you have going on that you could definitely test this idea. For the Airbnb example, this is most similar to your design, by showing the most highly used filters, then having the option to see or apply more with the "more filters" button

Take a look at those articles and let me know if you have any questions!

**Comment by corinne-nagel** [comment]
Hey all, I just saw this thread but I'm working [on this for Insights](https://www.figma.com/proto/lcBqy7Pvg0hBPPq50qRRe1/Insights---2022-Q1?node-id=4520%3A61946&scaling=min-zoom&page-id=4501%3A134800&starting-point-node-id=4520%3A61946&show-proto-sidebar=1). I signed up for office hours on the 2nd to discuss the state of this and the next steps. Would love to sync up on progress and steps forward so we don't duplicate work.

**Reply by yongjie1** [reply]
This is what our team is planning to build for the filter & search experience, here is the [design prototype](https://www.figma.com/proto/ozIl4CaeNYQkBpLYQ5u7kZ/Search-%26-Filter-on-call-list-page?page-id=227%3A70923&node-id=227%3A73710&viewport=299%2C48%2C0.33&scaling=scale-down&starting-point-node-id=227%3A73710&show-proto-sidebar=1). The idea is to show some commonly used filters as default, if users want to use more filters they can click on "more filters" button that will open up the side panel. 
I won't be able to make the office hour but let me know if you have any questions.

**Reply by corinne-nagel** [reply]
Ah I see, ok looks like you have a lot more filters you're working with. Have you had a chance to do user research you could share?

**Reply by yongjie1** [reply]
We've only had chance to chat with 2 clients, I won't call it a proper research. They are the business persons not the end users, and the chat is focused on the high level concept not the detailed interaction. In general, they know where to find the filters if they need more filters. When filters in the side panel has been applied, they think it's useful that we've showed the pills of the applied filters on the page. 
We've asked about frequently used filters but it seems that different types of job roles will have different frequently used filters. In future, we may allow users to set up the order of the filters based on their own needs.

**Reply by loreina** [reply]
Hi @corinne-nagel - just wanted to share the office hours recording with you in case you'd like to revisit any of our conversations. [Here's the video](https://www.loom.com/share/e37ae97022b543b78ba79a3e70e09f14) (it includes timestamps for each topic), and the password is posted in the #help-design-system channel. 

To summarize some of our conversation:
- Custom filtering won't be part of the filter group pattern for now, currently we have simple in scope for review with committee
- The designs you shared for custom filtering looks visually appealing, though we're unsure if it will account for other operators like AND/OR

Thanks again for coming to office hours!

**Comment by loreina** [comment, marked as answer]
We have a [Filter group pattern](https://paste.twilio.design/patterns/filter-group) available in Paste with guidance on simple table filtering. Marking this as answered due to inactivity.
